### PR TITLE
Add Swift

### PR DESCRIPTION
**Pull request by Cyberbeni**

### CODE REVIEW

**Comment by Cyberbeni** [comment]
I'd rather not commit 20 times to find what causes "Bad substitution", please fix it for me @asottile

**Comment by Cyberbeni** [comment]
Testing with swiftlint because that uses SourceKit which requires libsourcekitdInProc.so at /usr/lib/libsourcekitdInProc.so or LINUX_SOURCEKIT_LIB_PATH env var set to it's location. See: https://github.com/Cyberbeni/install-swift-tool/issues/68

**Comment by asottile** [comment]
> 
> 
> I'd rather not commit 20 times to find what causes "Bad substitution", please fix it for me @asottile

in dockerfiles the shell is `sh` which doesn't understand `${OS/.//}` -- probably a better way to do this would be to `. /etc/lsb-release && ` and then use the variables from that (here's what they look like):

```console
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.1 LTS"
```

**Comment by Cyberbeni** [comment]
There is still a spot where the dot is omitted from `DISTRIB_RELEASE`

**Comment by Cyberbeni** [comment]
Now the only question is how to run a tool that doesn't have a hooks.yaml? 🤔 The test should use https://github.com/realm/SwiftLint with `swiftlint lint --strict file`

**Comment by asottile** [comment]
seems like the pre-commit implementation of swift [doesn't yet support `additional_dependencies`](https://github.com/pre-commit/pre-commit/blob/2d54ea112aab5568c149bb81f428dce70010a151/pre_commit/languages/swift.py#L42) -- that would probably be the way

pre-commit currently tests the swift implementation by setting up a "remote repository" (on disk)

File: bin/test [hunk 1]
```diff
@@ -231,6 +231,52 @@ def test_can_run_rust_hook(tag: str, pc: str, podman: bool) -> None:
...
 
 
+SWIFT_HOOK = '''\
+repos:
+-   repo: https://github.com/nicklockwood/SwiftFormat
+    rev: 0.47.8
+    hooks:
+    -   id: swiftformat
+        name: Swiftformat
+        types: [swift]
```
Comment (asottile) [inline bin/test:241]: these lines don't do anything

**Comment by Cyberbeni** [comment]
(If you decide to split it to a different image, here is a list of required packages that I found after I got the libedit.so.2 missing error, as some of them were already included here: https://github.com/apple/swift-docker/blob/c0daa8c7a96e310e20c4b02bbfaf600f2d92eda9/5.3/ubuntu/20.04/Dockerfile )

I have a similarly good GitHub Action but that only uses the default GH Action caching, so rebuilding the tool happens more often (If activity in the repository is infrequent then this is a bit annoying), so I'm not in a hurry to get this merged.

**Comment by Cyberbeni** [comment]
Swift on Linux is not ABI stable yet, so if the swift version changes on the runner image, the tool has to be rebuilt, IDK if that is properly handled.

**Comment by asottile** [comment]
> 
> 
> Swift on Linux is not ABI stable yet, so if the swift version changes on the runner image, the tool has to be rebuilt, IDK if that is properly handled.

yes, pre-commit.ci assumes all languages are abi unstable and rebuilds on any image change

**Comment by asottile** [comment]
> a lot of these are redundant and/or install the world -- please don't add these packages unless there's actual demonstrable things that they break (and please include a test for that)

just making sure you didn't miss this

**Comment by Cyberbeni** [comment]
Ah, I overlooked that. I copied them from the official swift focal docker tag but since swiftformat currently is the only supported tool, I guess I'll test them 1-by-1 if they are required for it.

**Comment by Cyberbeni** [comment]
The `focal-slim` docker tag doesn't specify `libedit2` and `libz3-dev` but the test fails without them. The tests work correctly without `libcurl4` and `tzdata` buit they are included in `focal-slim`, is it okay to keep those 2 even though they doesn't seem to be required? (libcurl is probably used for networking and swiftformat doesn't use networking but there are other tools, like https://github.com/mono0926/LicensePlist that could be nice to integrate into pre-commit.ci but use networking)

**Comment by asottile** [comment]
this looks good, but I'm going to hold off on merging this for a bit -- I need to rethink how language support works for pre-commit.ci so that this image doesn't grow indefinitely (while still making it accessible for outsiders to run the image like CI does).

**Review by asottile — APPROVED** [review]

**Comment by asottile** [comment]
I've merged this via #69 -- it's not available yet (I need to adjust the backend to work with the new image format)

**Pull request closed without merging**